qiskit/pulse/builder.py

```diff
@@ -0,0 +1,1927 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+r"""Use the pulse builder to program pulse programs with an imperative
+assembly-like syntax.
+
+To begin pulse programming we must first initialize our program builder
+context with :func:`build`, after which we can begin
+adding program statements. For example, below we write a simple program that
+:func:`play`\s a pulse:
+
+.. jupyter-execute::
+
+    from qiskit import execute
+    from qiskit import pulse
+
+    d0 = pulse.DriveChannel(0)
+
+    with pulse.build() as pulse_prog:
+        pulse.play(pulse.Constant(100, 1.0), d0)
+
+    pulse_prog.draw()
+
+    # Execute on a real backend.
+    # job = execute(pulse_prog, backend)
+
+The builder initializes a :class:`pulse.Schedule`, ``pulse_prog``
+and then begins to construct the program within the context. The output pulse
+schedule will survive after the context is exited and can be executed like a
+normal Qiskit program ``qiskit.execute(pulse_prog, backend)``.
+
+Pulse programming has a simple imperative style. There is little need to deal
+with classes such as the :class:`~qiskit.circuit.QuantumCircuit` or
+:class:`pulse.Schedule`\s. The pulse builder dynamically constructs these
+intermediate representations (IR) behind the scenes. This leaves the programmer
+to worry about the raw experimental physics of pulse programming and not
+constructing cumbersome data structures.
+
+We can optionally pass a :class:`~qiskit.providers.BaseBackend` to
+:func:`build` to enable enhanced functionality. Below, we prepare a Bell state
+by automatically compiling the required pulses from their gate-level
+representations, while simultaneously applying a long decoupling pulse to a
+neighboring qubit. We terminate the experiment with a measurement to see which
+state we prepared. This program which mixes circuits and pulses will be
+automatically lowered to be run as a pulse program:
```

Comment:
```suggestion
state we prepared. This program which mixes circuits and pulses will be
automatically lowered to be run as a pulse program:
```
Similar comment as above